File: ios/Classes/FlEnvPlugin.swift
import Flutter
import UIKit

public class FlEnvPlugin: NSObject, FlutterPlugin {

    private var registry: [String: String] = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.pixmerc.fl_env/channel",
            binaryMessenger: registrar.messenger(),
        )
        let instance = FlEnvPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.initialize()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getValue":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(code: "INVALID_ARG", message: "Argument 'key' is required", details: nil))
                return
            }
            result(registry[key])

        case "getAll":
            result(registry)

        case "getActiveTier":
            result(RuntimeStorage.shared.getActiveTier() ?? "development")

        case "switchTier":
            result(FlutterError(
                code: "PHASE_RESTRICTION",
                message: "switchEnvironment is not supported in Phase 1.",
                details: nil,
            ))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize() {
        do {
            let key = try KeychainManager.getKey()
            guard let url = Bundle.main.url(forResource: "FlEnvRegistry", withExtension: "bin"),
                  let data = try? Data(contentsOf: url) else {
                NSLog("[FlEnv] FlEnvRegistry.bin not found in bundle — run 'fl_env build' first.")
                return
            }
            registry = try RegistryReader.readAll(key: key, from: data)
        } catch {
            NSLog("[FlEnv] initialization failed: %@", error.localizedDescription)
        }
    }
}

File: android/src/main/kotlin/com/pixmerc/fl_env/FlEnvPlugin.kt
package com.pixmerc.fl_env

import android.content.Context
import com.pixmerc.fl_env.keystore.KeyManager
import com.pixmerc.fl_env.registry.RegistryReader
import com.pixmerc.fl_env.storage.RuntimeStorage
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class FlEnvPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var storage: RuntimeStorage

    private var registry: Map<String, String> = emptyMap()
    private var initialized = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.pixmerc.fl_env/channel")
        channel.setMethodCallHandler(this)
        storage = RuntimeStorage(context)
        initialize()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getValue" -> {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARG", "Argument 'key' is required", null)
                } else {
                    result.success(registry[key])
                }
            }
            "getAll" -> result.success(registry)
            "getActiveTier" -> result.success(storage.getActiveTier() ?: "development")
            "switchTier" -> result.error(
                "PHASE_RESTRICTION",
                "switchEnvironment is not supported in Phase 1.",
                null,
            )
            else -> result.notImplemented()
        }
    }

    private fun initialize() {
        try {
            val key = KeyManager.getKey()
            registry = RegistryReader(context).readAll(key)
            initialized = true
        } catch (e: Exception) {
            // Log but do not crash — FlEnvService.init() will surface the error
            android.util.Log.e("FlEnv", "fl_env initialization failed: ${e.message}", e)
        }
    }
}
